lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java

```diff
@@ -0,0 +1,225 @@
+     * @return Itself to allow for  method chaining.
+     */
+    public TemporalEvaluator setDataSource(BinaryRatingDAO dao) {
+        dataSource = dao;
+        return this;
+    }
+
+    /**
+     * @param file The file contains the ratings input data
+     * @return itself
+     */
+    public TemporalEvaluator setDataSource(File file) throws IOException {
+        dataSource = BinaryRatingDAO.open(file);
+        return this;
+    }
+
+    /**
+     * @param file The file set as the output of the command
+     * @return Itself for  method chaining
+     */
+    public TemporalEvaluator setPredictOutputFile(File file) {
+        predictOutputFile = file;
+        return this;
+    }
+
+    /**
+     * *
+     *
+     * @param file  The file set as the output of the command
+     * @param cMode Compression Mode
+     * @return Itself for  method chaining
+     */
+
+    //TODO Set compression mode in file
+    public TemporalEvaluator setPredictOutputFile(File file, CompressionMode cMode) {
+        predictOutputFile = file;
+        return this;
+    }
+
+    /**
+     * @param time The time to rebuild
+     * @param unit Unit of time set
+     * @return Itself for  method chaining
+     */
+    public TemporalEvaluator setRebuildPeriod(Long time, TimeUnit unit) {
+        rebuildPeriod = unit.toSeconds(time);
+        return this;
+    }
+
+    /**
+     * @param seconds default rebuild period in seconds
+     * @return Itself for  method chaining
+     */
+    public TemporalEvaluator setRebuildPeriod(Long seconds) {
+        rebuildPeriod = seconds;
+        return this;
+    }
+
+    /**
+     * @return Returns prediction output file
+     */
+    public File getPredictOutputFile() {
+        return predictOutputFile;
+    }
+
+    /**
+     * @return Returns rebuild period
+     */
+    public Long getRebuildPeriod() {
+        return rebuildPeriod;
+    }
+
+    /**
+     * During the evaluation, it will replay the ratings, try to predict each one, and
+     * write the prediction, TARMSE and the rating to the output file
+     *
+     * @return Itself for  method chaining
+     */
+
+    public TemporalEvaluator execute() throws IOException, RecommenderBuildException {
+        Preconditions.checkState(algorithm != null, "no algorithm specified");
+        Preconditions.checkState(dataSource != null, "no input data specified");
+        Preconditions.checkState(predictOutputFile != null, "no output file specified");
+
+        TableWriter tableWriter;
+        //Builds file layout
+        TableLayoutBuilder tlb = new TableLayoutBuilder();
+        //file headers
+        tlb.addColumn("User")
+           .addColumn("Item")
+           .addColumn("Rating")
+           .addColumn("Timestamp")
+           .addColumn("Prediction")
+           .addColumn("TARMSE");
+
+        TableLayout tl = tlb.build();
+        tableWriter = CSVWriter.open(predictOutputFile, tl, CompressionMode.AUTO);
```

Comment:
This should be immediately before the outer `try`, ideally converted to a `try`-with-resources block.